saleor/account/tests/test_throttling.py

```diff
@@ -0,0 +1,387 @@
+import logging
+from datetime import timedelta
+from unittest.mock import patch
+
+import before_after
+import pytest
+from django.core.exceptions import ValidationError
+from django.utils import timezone
+from freezegun import freeze_time
+
+from ..error_codes import AccountErrorCode
+from ..throttling import (
+    MAX_DELAY,
+    MIN_DELAY,
+    authenticate_with_throttling,
+    get_cache_key_blocked_ip,
+    get_cache_key_failed_ip,
+    get_cache_key_failed_ip_with_user,
+    get_delay_time,
+)
+
+CORRECT_PASSWORD = "password"
+INCORRECT_PASSWORD = "incorrect-password"
+EXISTING_EMAIL = "test@example.com"
+NON_EXISTING_EMAIL = "non-existing@example.com"
+
+
+@pytest.mark.parametrize(
+    ("ip_attempts_count", "ip_user_attempts_count", "expected_delay"),
+    [
+        (1, 0, 1),
+        (0, 1, 1),
+        (21, 1, 4),
+        (21, 4, 8),
+        (100, 1, MAX_DELAY),
+        (1, 10, MAX_DELAY),
+        (0, 0, 1),
+        (-1, -17, 1),
+    ],
+)
+def test_get_delay_time(ip_attempts_count, ip_user_attempts_count, expected_delay):
+    # given & when
+    delay = get_delay_time(ip_attempts_count, ip_user_attempts_count)
+
+    # then
+    assert delay == expected_delay
+
+
+@freeze_time("2024-05-31 12:00:01")
+@patch("saleor.account.throttling.cache")
+def test_authenticate_successful_first_attempt(
+    mocked_cache, rf, customer_user, setup_mock_for_cache
+):
+    # given
+    dummy_cache = {}
+    setup_mock_for_cache(dummy_cache, mocked_cache)
+
+    ip = "123.123.123.123"
+    request = rf.request(HTTP_X_FORWARDED_FOR=ip)
+
+    block_key = get_cache_key_blocked_ip(ip)
+    ip_key = get_cache_key_failed_ip(ip)
+    ip_user_key = get_cache_key_failed_ip_with_user(ip, customer_user.id)
+
+    # when
+    user = authenticate_with_throttling(request, EXISTING_EMAIL, CORRECT_PASSWORD)
+
+    # then
+    assert user
+    assert mocked_cache.get(block_key) is None
```

Comment:
If you expect that this values will not be set by your logic, then would be good to make an explicit assert:
`assert not mocked.set.called`
This will confirm that you didn't call any `set` in your logic. 
If you expect some `set` calls, but you want to check that the `set` was not called for `block_key`, you could do something like:
`assert block_key not in mocked_cache.set.call_args_list`

This makes your test more explicit, as I will already know what should happen with the cache. Currently I know that these fields are None, but I don't know if they are None, because you didn't call `.set` for them or because you call `set` with None values

Reply:
Perhaps we could make this test more explicit by using `assert mocked_cache.mock_calls = [...]`? This would verify order of the calls and what was passed to each call thus any change to order of calls or passed args would make this test fail

Reply:
More assertions were added.